src/sidebar/oauth-auth.js

```diff
@@ -16,23 +16,62 @@ function auth($http, settings) {
   var accessTokenPromise;
   var tokenUrl = resolve('token', settings.apiUrl);
 
+  // Return a new tokenInfo object from the given tokenUrl endpoint response.
+  function tokenInfoFrom(response) {
+    var data = response.data;
+    return {
+      accessToken:  data.access_token,
+      expiresIn:    data.expires_in,
+      refreshToken: data.refresh_token,
+    };
+  }
+
+  // Post the given data to the tokenUrl endpoint as a form submission.
+  // Return a Promise for the POST request.
+  function postToTokenUrl(data) {
+    data = queryString.stringify(data);
+    var requestConfig = {
+      headers: {'Content-Type': 'application/x-www-form-urlencoded'},
+    };
+    return $http.post(tokenUrl, data, requestConfig);
+  }
+
   // Exchange the JWT grant token for an access token.
   // See https://tools.ietf.org/html/rfc7523#section-4
   function exchangeToken(grantToken) {
-    var data = queryString.stringify({
+    var data = {
       grant_type: 'urn:ietf:params:oauth:grant-type:jwt-bearer',
       assertion: grantToken,
-    });
-    var requestConfig = {
-      headers: {'Content-Type': 'application/x-www-form-urlencoded'},
     };
-    return $http.post(tokenUrl, data, requestConfig)
-      .then(function (response) {
-        if (response.status !== 200) {
-          throw new Error('Failed to retrieve access token');
-        }
-        return response.data;
-      });
+    return postToTokenUrl(data).then(function (response) {
+      if (response.status !== 200) {
+        throw new Error('Failed to retrieve access token');
+      }
+      return tokenInfoFrom(response);
+    });
+  }
+
+  // Exchange the refresh token for a new access token and refresh token pair.
+  // See https://tools.ietf.org/html/rfc6749#section-6
+  function refreshAccessToken(refreshToken) {
+    var data = {grant_type: 'refresh_token', refresh_token: refreshToken};
+    postToTokenUrl(data).then(function (response) {
```

Comment:
Yeah, I know it will be, but oauth error handling in the client is generally pretty nonexistent, so I want to do a separate PR (probably multiple PRs) for it.